File: backend/app/models/shopping.py
from pydantic import BaseModel, Field
from .credit_card import CreditCardInput
from .store_rewards import StoreRewardsProgram
from .product import Store


class ShoppingItem(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    quantity: int = Field(ge=1, le=99, default=1)
    unit: str | None = None


class ComparisonRequest(BaseModel):
    items: list[ShoppingItem] = Field(min_length=1, max_length=50)
    zip_code: str = Field(pattern=r"^\d{5}$")
    credit_cards: list[CreditCardInput] = []
    store_rewards_programs: list[StoreRewardsProgram] = []
    demo_mode: bool = False


class LocationResult(BaseModel):
    store: Store
    store_id: str
    name: str
    address: str
    distance_miles: float | None = None

File: frontend/src/api/shopping.ts
import client from './client'
import type { ComparisonResult } from '../types/product'
import type { ShoppingItem } from '../types/shopping-list'
import type { CreditCard } from '../types/credit-card'
import type { StoreRewardsProgram } from '../types/store-rewards'

export async function fetchComparison(
  items: ShoppingItem[],
  zipCode: string,
  creditCards: CreditCard[],
  storeRewardsPrograms: StoreRewardsProgram[],
  demoMode = false,
): Promise<ComparisonResult> {
  const payload = {
    items: items.map(({ name, quantity, unit }) => ({ name, quantity, unit })),
    zip_code: zipCode,
    credit_cards: creditCards,
    store_rewards_programs: storeRewardsPrograms,
    demo_mode: demoMode,
  }
  const res = await client.post<ComparisonResult>('/compare', payload)
  return res.data
}
